packages/deployer/src/compiler.ts

```diff
@@ -44,123 +49,59 @@ const SOLC_BIN_DIR = path.join(__dirname, '..', '..', 'solc_bin');
      */
     constructor(opts: CompilerOptions) {
-        this._contractDirs = opts.contractDirs;
+        this._contractsDir = opts.contractsDir;
         this._networkId = opts.networkId;
         this._optimizerEnabled = opts.optimizerEnabled;
         this._artifactsDir = opts.artifactsDir;
         this._specifiedContracts = opts.specifiedContracts;
+        this._nameResolver = new NameResolver(path.resolve(this._contractsDir));
+        const resolver = new FallthroughResolver();
+        resolver.appendResolver(new URLResolver());
+        resolver.appendResolver(new NPMResolver(path.resolve('')));
+        resolver.appendResolver(new FSResolver());
+        resolver.appendResolver(this._nameResolver);
+        this._resolver = resolver;
     }
     /**
      * Compiles selected Solidity files found in `contractsDir` and writes JSON artifacts to `artifactsDir`.
      */
     public async compileAsync(): Promise<void> {
         await createDirIfDoesNotExistAsync(this._artifactsDir);
         await createDirIfDoesNotExistAsync(SOLC_BIN_DIR);
-        this._contractSources = {};
-        const contractIdToSourceFileId: ContractIdToSourceFileId = {};
-        const contractDirs = Array.from(this._contractDirs.values());
-        for (const contractDir of contractDirs) {
-            const sources = await Compiler._getContractSourcesAsync(contractDir.path, contractDir.path);
-            _.forIn(sources, (source, sourceFilePath) => {
-                const sourceFileId = constructUniqueSourceFileId(contractDir.namespace, sourceFilePath);
-                // Record the file's source and data
-                if (!_.isUndefined(this._contractSources[sourceFileId])) {
-                    throw new Error(`Found duplicate source files with ID '${sourceFileId}'`);
-                }
-                this._contractSources[sourceFileId] = source;
-                // Create a mapping between the contract id and its source file id
-                const contractId = constructContractId(contractDir.namespace, sourceFilePath);
-                if (!_.isUndefined(contractIdToSourceFileId[contractId])) {
-                    throw new Error(`Found duplicate contract with ID '${contractId}'`);
-                }
-                contractIdToSourceFileId[contractId] = sourceFileId;
-            });
+        let contractNamesToCompile: string[] = [];
+        if (this._specifiedContracts.has(ALL_CONTRACTS_IDENTIFIER)) {
+            const allContracts = this._nameResolver.getAllContracts();
+            contractNamesToCompile = _.map(allContracts, contractSource =>
+                path.basename(contractSource.path, constants.SOLIDITY_FILE_EXTENSION),
+            );
+        } else {
+            contractNamesToCompile = Array.from(this._specifiedContracts.values());
+        }
+        for (const contractNameToCompile of contractNamesToCompile) {
+            await this._compileContractAsync(contractNameToCompile);
```

Comment:
It could. The only problem is that we lazy-load the compiler versions. If we run in parallel - we should either do some locking/cond vars or double-load some compilers. I've tried that before and the speed gain wasn't noticeable.